CSharp/JobPrep/Program.cs

```diff
@@ -0,0 +1,333 @@
+                IEnumerable<ICloudWorkItem> wis = wm.ListWorkItems();
+                foreach (var w in wis)
+                {
+                    Console.WriteLine("Workitem: " + w.Name + " State:" + w.State);
+                }
+                Console.WriteLine();
+            }
+        }
+
+        private static async Task<bool> PollJobPrepTaskAsync(IPoolManager pm, string pool, string vm, string workitem, string job, int sleep)
+        {
+            while(true)
+            {
+                try
+                {
+                    Console.WriteLine("polling prep task on " + vm);
+                    var preptask = await pm.GetJobPreparationTaskStatusAsync(pool, vm, workitem, job);
+                    if (preptask.State == JobPreparationTaskState.Completed)
+                        return preptask.ExitCode == 0;
+                    Console.WriteLine("sleep a sec");
+                    await Task.Delay(sleep);
+                }
+                catch (AggregateException) { // ignore not found exception
+                    Console.WriteLine("caught exception in async method");
+                }
+            }
+        }
+
+        private static void Log(string msg, string vm = "")
+        {
+#if VERBOSE
+            Console.WriteLine(DateTime.Now.ToString("HH:mm:ss.ffff ") + vm + " " + msg);
+#endif
+        }
+
+        private static bool WaitAllJobPrepTask(IBatchClient client, string pool, string workitem, string job, Func<int, int, bool> stop, Func<int, int, bool> succeed, int timeout = 60*1000, int freq = 2000, int parallelism = 5)
+        {
+            using (IPoolManager pm = client.OpenPoolManager())
+            {
+                using (IWorkItemManager wm = client.OpenWorkItemManager())
+                {
+
+                    CancellationTokenSource cts = new CancellationTokenSource();
+                    CancellationToken ct = cts.Token;
+
+                    BlockingCollection<string> vms = new BlockingCollection<string>();
+                    List<Task> tasks = new List<Task>();
+                    int nb_succeeded = 0;
+                    int nb_failed = 0;
+
+                    // find all VM in the Pool and put them in a queue
+                    Log("add vms");
+                    tasks.Add(Task.Run(async () =>
+                    {
+                        Log("start adding vms");
+                        // list all vms and enqueue
+                        ICloudPool p = pm.GetPool(PoolName);
+                        var en = p.ListVMs(new ODATADetailLevel(selectClause: "name")).GetAsyncEnumerator();
+                        while(await en.MoveNextAsync())
+                        {
+                            Log("add vm", en.Current.Name);
+                            vms.Add(en.Current.Name);
+                            if (ct.IsCancellationRequested)
+                                break;
+                        }
+                    }));
```

Comment:
Why is all of this happening in a task.run?  Can't you just make this whole method async and then drop the task.run part and do the await in the actual method?